### PR TITLE
Consolidate logic for resolving hub

### LINKED ISSUE #787: Sentry crashes when SENTRY_DSN is defined
### Environment

How do you use Sentry?
Sentry SaaS (sentry.io) or self-hosted/on-premise (which version?)

_**We use Sentry SaaS.**_

Which SDK and version?
e.g: JavaScript 5.11.1, .NET 1.2.0

_**.NET Core 3.1 and .NET Core 5.0. Sentry.AspNetCore 3.0.2**_

### Steps to Reproduce

https://github.com/aleksihakli/sentry-error

Run `SENTRY_DSN="https://a@b.ingest.sentry.io/c" dotnet run` command with `Sentry.AspNetCore` version `3.0.2` which crashes the program.

Running the same command with `Sentry.AspNetCore` version `2.1.8` does not crash.

### Expected Result

_**.NET Core and Sentry run normally.**_

### Actual Result

What actually happened. Maybe a screenshot/recording? Maybe some logs?

_**Sentry crashes the .NET Core program when `SENTRY_DSN` environment variable is set:**_

I have tried to define valid and invalid keys with `set`, `export` and direct definition and they all crash.

```
$ SENTRY_DSN="https://a@b.ingest.sentry.io/c" dotnet run
Building...
Unhandled exception. System.ArgumentException: No DSN defined in the SentryOptions
   at Sentry.Internal.SdkComposer..ctor(SentryOptions options) in /_/src/Sentry/Internal/SdkComposer.cs:line 15
   at Sentry.SentryClient..ctor(SentryOptions options, IBackgroundWorker worker) in /_/src/Sentry/SentryClient.cs:line 53
   at Sentry.SentryClient..ctor(SentryOptions options) in /_/src/Sentry/SentryClient.cs:line 41
   at Sentry.Internal.Hub..ctor(SentryOptions options) in /_/src/Sentry/Internal/Hub.cs:line 62
   at Sentry.Internal.OptionalHub.FromOptions(SentryOptions options) in /_/src/Sentry/Internal/OptionalHub.cs:line 23
   at Sentry.Extensions.Logging.Extensions.DependencyInjection.ServiceCollectionExtensions.<>c__0`1.<AddSentry>b__0_3(IServiceProvider c) in /_/src/Sentry.Extensions.Logging/Extensions/DependencyInjection/ServiceCollectionExtensions.cs:line 37
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteRuntimeResolver.VisitFactory(FactoryCallSite factoryCallSite, RuntimeResolverContext context)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteVisitor`2.VisitCallSiteMain(ServiceCallSite callSite, TArgument argument)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteRuntimeResolver.VisitCache(ServiceCallSite callSite, RuntimeResolverContext context, ServiceProviderEngineScope serviceProviderEngine, RuntimeResolverLock lockType)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteRuntimeResolver.VisitRootCache(ServiceCallSite singletonCallSite, RuntimeResolverContext context)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteVisitor`2.VisitCallSite(ServiceCallSite callSite, TArgument argument)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteRuntimeResolver.Resolve(ServiceCallSite callSite, ServiceProviderEngineScope scope)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.DynamicServiceProviderEngine.<>c__DisplayClass1_0.<RealizeService>b__0(ServiceProviderEngineScope scope)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.ServiceProviderEngine.GetService(Type serviceType, ServiceProviderEngineScope serviceProviderEngineScope)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.ServiceProviderEngineScope.GetService(Type serviceType)
   at Microsoft.Extensions.DependencyInjection.ServiceProviderServiceExtensions.GetRequiredService(IServiceProvider provider, Type serviceType)
   at Microsoft.Extensions.DependencyInjection.ServiceProviderServiceExtensions.GetRequiredService[T](IServiceProvider provider)
   at Sentry.Extensions.Logging.Extensions.DependencyInjection.ServiceCollectionExtensions.<>c__0`1.<AddSentry>b__0_2(IServiceProvider c) in /_/src/Sentry.Extensions.Logging/Extensions/DependencyInjection/ServiceCollectionExtensions.cs:line 29
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteRuntimeResolver.VisitFactory(FactoryCallSite factoryCallSite, RuntimeResolverContext context)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteVisitor`2.VisitCallSiteMain(ServiceCallSite callSite, TArgument argument)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteRuntimeResolver.VisitDisposeCache(ServiceCallSite transientCallSite, RuntimeResolverContext context)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteVisitor`2.VisitCallSite(ServiceCallSite callSite, TArgument argument)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteRuntimeResolver.VisitConstructor(ConstructorCallSite constructorCallSite, RuntimeResolverContext context)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteVisitor`2.VisitCallSiteMain(ServiceCallSite callSite, TArgument argument)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteRuntimeResolver.VisitCache(ServiceCallSite callSite, RuntimeResolverContext context, ServiceProviderEngineScope serviceProviderEngine, RuntimeResolverLock lockType)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteRuntimeResolver.VisitRootCache(ServiceCallSite singletonCallSite, RuntimeResolverContext context)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteVisitor`2.VisitCallSite(ServiceCallSite callSite, TArgument argument)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteRuntimeResolver.VisitIEnumerable(IEnumerableCallSite enumerableCallSite, RuntimeResolverContext context)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteVisitor`2.VisitCallSiteMain(ServiceCallSite callSite, TArgument argument)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteRuntimeResolver.VisitCache(ServiceCallSite callSite, RuntimeResolverContext context, ServiceProviderEngineScope serviceProviderEngine, RuntimeResolverLock lockType)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteRuntimeResolver.VisitRootCache(ServiceCallSite singletonCallSite, RuntimeResolverContext context)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteVisitor`2.VisitCallSite(ServiceCallSite callSite, TArgument argument)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteRuntimeResolver.VisitConstructor(ConstructorCallSite constructorCallSite, RuntimeResolverContext context)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteVisitor`2.VisitCallSiteMain(ServiceCallSite callSite, TArgument argument)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteRuntimeResolver.VisitCache(ServiceCallSite callSite, RuntimeResolverContext context, ServiceProviderEngineScope serviceProviderEngine, RuntimeResolverLock lockType)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteRuntimeResolver.VisitRootCache(ServiceCallSite singletonCallSite, RuntimeResolverContext context)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteVisitor`2.VisitCallSite(ServiceCallSite callSite, TArgument argument)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteRuntimeResolver.VisitConstructor(ConstructorCallSite constructorCallSite, RuntimeResolverContext context)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteVisitor`2.VisitCallSiteMain(ServiceCallSite callSite, TArgument argument)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteRuntimeResolver.VisitCache(ServiceCallSite callSite, RuntimeResolverContext context, ServiceProviderEngineScope serviceProviderEngine, RuntimeResolverLock lockType)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteRuntimeResolver.VisitRootCache(ServiceCallSite singletonCallSite, RuntimeResolverContext context)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteVisitor`2.VisitCallSite(ServiceCallSite callSite, TArgument argument)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteRuntimeResolver.VisitConstructor(ConstructorCallSite constructorCallSite, RuntimeResolverContext context)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteVisitor`2.VisitCallSiteMain(ServiceCallSite callSite, TArgument argument)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteRuntimeResolver.VisitCache(ServiceCallSite callSite, RuntimeResolverContext context, ServiceProviderEngineScope serviceProviderEngine, RuntimeResolverLock lockType)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteRuntimeResolver.VisitRootCache(ServiceCallSite singletonCallSite, RuntimeResolverContext context)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteVisitor`2.VisitCallSite(ServiceCallSite callSite, TArgument argument)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteRuntimeResolver.VisitConstructor(ConstructorCallSite constructorCallSite, RuntimeResolverContext context)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteVisitor`2.VisitCallSiteMain(ServiceCallSite callSite, TArgument argument)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteRuntimeResolver.VisitCache(ServiceCallSite callSite, RuntimeResolverContext context, ServiceProviderEngineScope serviceProviderEngine, RuntimeResolverLock lockType)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteRuntimeResolver.VisitRootCache(ServiceCallSite singletonCallSite, RuntimeResolverContext context)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteVisitor`2.VisitCallSite(ServiceCallSite callSite, TArgument argument)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteRuntimeResolver.Resolve(ServiceCallSite callSite, ServiceProviderEngineScope scope)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.DynamicServiceProviderEngine.<>c__DisplayClass1_0.<RealizeService>b__0(ServiceProviderEngineScope scope)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.ServiceProviderEngine.GetService(Type serviceType, ServiceProviderEngineScope serviceProviderEngineScope)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.ServiceProviderEngine.GetService(Type serviceType)
   at Microsoft.Extensions.DependencyInjection.ServiceProvider.GetService(Type serviceType)
   at Microsoft.Extensions.DependencyInjection.ServiceProviderServiceExtensions.GetRequiredService(IServiceProvider provider, Type serviceType)
   at Microsoft.Extensions.DependencyInjection.ServiceProviderServiceExtensions.GetRequiredService[T](IServiceProvider provider)
   at Microsoft.Extensions.Hosting.HostBuilder.Build()
   at sentry_error.Program.Main(String[] args) in /private/tmp/sentry-error/Program.cs:line 16
```

### PR DESCRIPTION
**Pull request by Tyrrrz**
Fixing this reminded me that `SentryOptions.Dsn` is a hot mess and we should really refactor it.

Fixes #787

### CODE REVIEW

File: test/Sentry.Tests/SentrySdkTests.cs [hunk 1]
```diff
@@ -95,17 +95,19 @@ public void Init_CallbackWithoutDsn_ValidDsnEnvironmentVariable_LocatesDsnEnviro
         }
 
         [Fact]
-        public void Init_CallbackWithoutDsn_InvalidDsnEnvironmentVariable_DisabledSdk()
```
Comment (Tyrrrz) [inline test/Sentry.Tests/SentrySdkTests.cs:98]: No idea why this expected it to return `DisabledSdk` considering that a similar test expects an exception: `Init_InvalidDsnEnvironmentVariable_Throws`. I believe throwing an exception is the way to go here.

File: src/Sentry.AspNetCore/ApplicationBuilderExtensions.cs [hunk 2]
```diff
@@ -44,7 +44,7 @@ public static IApplicationBuilder UseSentry(this IApplicationBuilder app)
                 var stackTraceFactory = app.ApplicationServices.GetService<ISentryStackTraceFactory>();
                 if (stackTraceFactory != null)
                 {
-                    _ = o.UseStackTraceFactory(stackTraceFactory);
+                    o.UseStackTraceFactory(stackTraceFactory);
```
Comment (bruno-garcia) [inline src/Sentry.AspNetCore/ApplicationBuilderExtensions.cs:47]: WHat about a find and replace on the whole project if this will be a thing now?
Better than noise on each PR.

Also editorconfig rule to warn to add this has to go if it's not yet gone

File: src/Sentry/SentrySdk.cs [hunk 3]
```diff
@@ -28,13 +28,39 @@ public static class SentrySdk
         /// </summary>
         public static SentryId LastEventId { [DebuggerStepThrough] get => _hub.LastEventId; }
 
+        internal static IHub InitHub(SentryOptions options)
+        {
+            // Side-effects in a factory function 🤮
```
Comment (bruno-garcia) [inline src/Sentry/SentrySdk.cs:33]: Suggestion for an alternative approach? We want to setup logging before anything runs
```suggestion
```

File: src/Sentry/SentrySdk.cs [hunk 4]
```diff
@@ -28,13 +28,39 @@ public static class SentrySdk
...
+            {
+                options.DiagnosticLogger?.LogWarning(
+                    "Init was called but no DSN was provided nor located. Sentry SDK will be disabled."
+                );
+
+                return DisabledHub.Instance;
+            }
+
+            // Validate DSN for an early exception in case it's malformed
+            _ = Dsn.Parse(dsn);
```
Comment (bruno-garcia) [inline src/Sentry/SentrySdk.cs:52]: When to use the discard operator vs when not to?

In this PR you've removed from a code unrelated to the changed and added here

File: test/Sentry.Tests/SentrySdkTests.cs (hunk 1 above)
Comment (bruno-garcia) [reply test/Sentry.Tests/SentrySdkTests.cs:98]: Invalid DSN is a clear mistake so makes sense to throw.

I wouldn't disagree that if `Dsn` is `null` we can throw (only "" is disable) at init but this is a breaking change

File: src/Sentry/SentrySdk.cs (hunk 4 above)
Comment (Tyrrrz) [reply src/Sentry/SentrySdk.cs:52]: Here the discard is used to signal that the return of an otherwise pure function is ignored and it's only called for the exception (which is debatably a side-effect, but it's usually not considered as one). Basically, in 99.9% of cases you will want to have the return value, but this is that 0.01% case where we don't. The discard is meant to highlight that the return value (which is the most element part of this function) is ignored on purpose and is not a mistake (albeit a hack).

In other cases, like in case of `o.UseStackTraceFactory` the function is inherently effectful and it's called for its side-effects rather than the return value. Conversely to the other case, the return of `UseStackTraceFactory` is only ever needed in 0.01% of cases and then only for convenience. Here, ignoring the return value is absolutely normal.

So, in short:

Use discard to highlight that a return method of an (almost) pure function is discarded on purpose.

Don't use discard when ignoring return of an inherently effectful method, because the main point of executing said method is in the side-effects and not the return value.

File: src/Sentry/SentrySdk.cs (hunk 3 above)
Comment (Tyrrrz) [reply src/Sentry/SentrySdk.cs:33]: For example, instead of mutating `SentryOptions` here, we could have `SentryOptions.GetLoggerOrDefault()` that would return the user-configured logger or initialize a console logger. Essentially, change the behavior from push-based to pull-based.

Comment (Tyrrrz) [reply src/Sentry/SentrySdk.cs:33]: Although I concede that is rather nitpicky at this point because there are many other places where `SentryOptions` is mutated.

File: src/Sentry.AspNetCore/ApplicationBuilderExtensions.cs (hunk 2 above)
Comment (Tyrrrz) [reply src/Sentry.AspNetCore/ApplicationBuilderExtensions.cs:47]: I have never had a warning about this, so I presume there is no rule in editorconfig, or it's just not enforced.
I will search & replace in another PR.

File: test/Sentry.Tests/SentrySdkTests.cs (hunk 1 above)
Comment (Tyrrrz) [reply test/Sentry.Tests/SentrySdkTests.cs:98]: I think the key point here is that (in the current state) when the user sets `SENTRY_DSN` environment variable to an invalid dsn, it will just be ignored. I think if that environment variable is set, it was a conscious decision, no less than setting it in `appsettings.json`, so it should throw an exception instead.

**Comment by Tyrrrz** [comment]
@bruno-garcia good to merge?

File: src/Sentry/SentrySdk.cs [hunk 5]
```diff
@@ -28,13 +28,39 @@ public static class SentrySdk
...
         public static SentryId LastEventId { [DebuggerStepThrough] get => _hub.LastEventId; }
 
+        internal static IHub InitHub(SentryOptions options)
+        {
+            // Side-effects in a factory function 🤮
+            options.SetupLogging();
+
+            // If DSN is null (i.e. not explicitly disabled, just unset), then
+            // try to resolve the value from environment.
+            var dsn = options.Dsn ??= DsnLocator.FindDsnStringOrDisable();
```
Comment (aleksihakli) [inline src/Sentry/SentrySdk.cs:38]: 👍

**Comment by aleksihakli** [comment]
IMO looks clean, would be awesome to get this into a patch release so we can test it in our stack and get performance tracing going for .NET.

File: src/Sentry.AspNetCore/ApplicationBuilderExtensions.cs (hunk 2 above)
Comment (bruno-garcia) [reply src/Sentry.AspNetCore/ApplicationBuilderExtensions.cs:47]: https://github.com/getsentry/sentry-dotnet/blob/54af283fa2c788e5416c23dbba5f55c6698c5b94/.editorconfig#L120

File: src/Sentry/SentrySdk.cs (hunk 3 above)
Comment (bruno-garcia) [reply src/Sentry/SentrySdk.cs:33]: The goal is to set an instance on the options so it's available throughout the SDK that's relying on `options.Logger?.LogX`

File: src/Sentry/SentrySdk.cs (hunk 4 above)
Comment (bruno-garcia) [reply src/Sentry/SentrySdk.cs:52]: >  The discard is meant to highlight that the return value (which is the most important element of this function) is ignored on purpose and is not a mistake (albeit a hack).

And that's why I enabled warning when a return type isn't assigned or explicitly discarded.

> Don't use discard when ignoring return of an inherently effectful method, because the main point of executing said method is in the side-effects and not the return value.

I'm fine with that but please make the change in 1 PR to avoid the noise of this refactor in every single change

File: test/Sentry.Tests/SentrySdkTests.cs (hunk 1 above)
Comment (bruno-garcia) [reply test/Sentry.Tests/SentrySdkTests.cs:98]: Makes sense.

**Review by bruno-garcia — APPROVED** [review]
LGTM

**Comment by bruno-garcia** [comment]
@aleksihakli we'll make a release soon (this week for sure)

**Comment by bruno-garcia** [comment]
@aleksihakli it's out 3.0.5

**Comment by aleksihakli** [comment]
@bruno-garcia I'll get to testing it today, thank you!

**Comment by aleksihakli** [comment]
Seems to work, digging through the .NET tracing sample rates and other configurations next :) Thanks @Tyrrrz and @bruno-garcia!
